pkg/util/validation/limits.go

```diff
@@ -253,13 +253,24 @@ func SetDefaultLimitsForYAMLUnmarshalling(defaults Limits) {
 	defaultLimits = &defaults
 }
 
+type ChangeListener func()
+type ChangeListenerRef interface{}
+
 // TenantLimits exposes per-tenant limit overrides to various resource usage limits
 type TenantLimits interface {
 	// ByUserID gets limits specific to a particular tenant or nil if there are none
 	ByUserID(userID string) *Limits
 
 	// AllByUserID gets a mapping of all tenant IDs and limits for that user
 	AllByUserID() map[string]*Limits
+
+	// AddChangeListener adds a listener function which will get called each time the tenants
```

Comment:
I'm a little concerned about adding this behavior to the `TenantLimits` interface. GEM has a handful of implementations of it and it's not clear to me how they would implement this. For example, one implementation wraps a multiple other implementations to allow multiple sources for per-tenant overrides.

Reply:
Good point. Will look into it before merging this PR.

Reply:
I took a look at GEM. There are 2 implementations of `TenantLimits` and I think adding support there is totally doable, just some extract code to write.

An alternative could be remove the `ChangeListener` from `TenantLimits` and just keep it in `Overrides`, then have a background goroutine periodically polling `TenantLimits` to check if they changed. The reason why we can't move this polling to store-gateway itself is because `Overrides` doesn't export any way to get "all tenants limits" or the "list of tenants for which an override has been set". I don't like much this option because having to deal with the goroutine in `Overrides` make be annoying.

Another alternative, could be adding a function "list of tenants for which an override has been set" to `Overrides`. With that function then we could implement a polling directly in the store-gateway. The store-gateway would have to list all tenants first, and then call `StoreGatewayTenantShardSize()` for each of them.

Finally, one last alternative that comes to my mind is adding a function `Overrides.StoreGatewayTenantShardSize() map[string]int` returning the `StoreGatewayTenantShardSize` setting for all users. The store-gateway would then just poll it. This implementation would be definitely easier and probably cleaner. The only drawback is that the store-gateway would do its own polling (at its own hardcoded period) instead of being notified as soon as the config has been reloaded and changed.

Reply:
Personally I would prefer a solution where `validation` package doesn't need to be modified. Store-gateway just needs some notifications when it needs to do a resync. It can get these notifications from a channel injected via constructor. If channel is nil (eg. no runtime manager is running), no notifications are delivered. This design would keep `TenantLimits` unchanged, would not require no "empty" types in `validation` package, and it would make the dependency clear (explicit in the constructor).